proxy/coco/src/peer.rs

```diff
@@ -123,385 +121,428 @@ impl Peer {
-                },
-                Err(err) => {
-                    log::error!("announce error: {:?}", err);
-                    sender.send(AnnounceEvent::Failed).await.ok();
-                },
+impl Future for Subroutines {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut events = Vec::with_capacity(9);
+
+        // FIXME(kim): can we not somehow coalesce all of these into a stream?
+        {
+            if let Poll::Ready(Some(protocol_event)) = self.protocol_events.poll_next_unpin(cx) {
+                events.push(Event::Protocol(protocol_event));
             }
-        };
-        tokio::pin!(go);
 
-        future::select(stop, go).map(|_| ()).await;
-    });
+            if let Poll::Ready(Some(_)) = self.announce_timer.poll_next_unpin(cx) {
+                events.push(Event::Announce(AnnounceEvent::Tick));
+            }
+
+            if let Poll::Ready(Some(_)) = self.ping_timer.poll_next_unpin(cx) {
+                events.push(Event::Ping);
+            }
+
+            if let Poll::Ready(Some(announce_event)) = self.announcements.poll_recv(cx) {
+                events.push(Event::Announce(announce_event));
+            }
+
+            if let Poll::Ready(Some(sync_event)) = self.peer_syncs.poll_recv(cx) {
+                events.push(Event::PeerSync(sync_event));
+            }
+
+            if let Poll::Ready(Some(timeout_event)) = self.timeouts.poll_recv(cx) {
+                events.push(Event::Timeout(timeout_event));
+            }
+
+            if let Poll::Ready(Some(urn)) = self.request_queries.poll_next_unpin(cx) {
+                events.push(Event::Request(RequestEvent::Query(urn)));
+            }
+
+            if let Poll::Ready(Some(url)) = self.request_clones.poll_next_unpin(cx) {
+                events.push(Event::Request(RequestEvent::Clone(url)));
+            }
+
+            if let Poll::Ready(Some(request_event)) = self.requests.poll_next_unpin(cx) {
+                events.push(Event::Request(request_event));
+            }
+        }
+
+        for event in events {
+            log::debug!("handling subroutine event: {:?}", event);
+            self.subscriber.send(event.clone()).ok();
+            for cmd in self.run_state.transition(&event) {
+                let task = match cmd {
+                    Command::Announce => SpawnAbortable::new(announce(
+                        self.state.clone(),
+                        self.store.clone(),
+                        self.announce_sender.clone(),
+                    )),
+                    Command::SyncPeer(peer_id) => SpawnAbortable::new(sync(
+                        self.state.clone(),
+                        peer_id.clone(),
+                        self.peer_sync_sender.clone(),
+                    )),
+                    Command::StartSyncTimeout(sync_period) => SpawnAbortable::new(
```

Comment:
The semantics of this are a bit unclear: if the `sync` task is supposed to be cancelled when this timer elapses, we need to schedule both at the same time -- since everything here is concurrent (parallel as Rustaeans prefer to call it), we don't actually know _which_ `sync` task we're supposed to cancel. In this case, `RunState` should not be concerned with the timeout.

Reply:
/cc @xla @FintanH

Reply:
The sync timeout is not so much to abort ongoing sync operations, but to guarantee that the peer goes "online", so that subroutines which depend on the peer being in that state can start operating. So the sync timeout is global for the entire peer. If we want explicit timeout/cancellation per sync task it should be solved in isolation.

Reply:
Ok fair